File: 56.merge-intervals.js
/*
 * @lc app=leetcode id=56 lang=javascript
 *
 * [56] Merge Intervals
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var merge = function(intervals) {
  // Sort, then merge overlapping intervals
  // O(nlog(n)) time
  // O(k) space
  intervals.sort((a, b) => a[0] - b[0])
  const withinInterval = (arrA, arrB) => {
    // check if minimum of arrB is
    // less than or equal to max of arrA
    if (arrB[0] <= arrA[1]) return true
    return false
  }
  let index = 0
  while (index < intervals.length -1) {
    if (withinInterval(intervals[index], intervals[index+1])) {
      // merge
      let min = intervals[index][0]
      let max = Math.max(intervals[index][1], intervals[index+1][1])
      intervals.splice(index, 2, [min, max])
    } else {
      index++
    }
  }
  return intervals
};
// @lc code=end

const intervals = [[1,3],[2,6],[15,18], [8,10]]
console.log(merge(intervals))

File: cpp/56.merge-intervals.cpp
/*
 * @lc app=leetcode id=56 lang=cpp
 *
 * [56] Merge Intervals
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> merge(vector<vector<int>> &intervals)
    {
        // for each interval if the prev interval is intersecting
        // the curr interval, then merge
        // O(nlog(n) + n) time O(n) space
        sort(intervals.begin(), intervals.end(), [](vector<int> const &a, vector<int> const &b)
             { return a[0] < b[0]; });
        vector<vector<int>> mergeIntervals;
        mergeIntervals.push_back(intervals[0]);
        for (int i = 0; i < intervals.size(); i++)
        {
            if (mergeIntervals.back()[1] >= intervals[i][0])
            {
                mergeIntervals.back()[1] = max(mergeIntervals.back()[1], intervals[i][1]);
            }
            else
            {
                mergeIntervals.push_back(intervals[i]);
            }
        }
        return mergeIntervals;
    }
};
// @lc code=end
